File: item-list/src/main/java/com/gurunars/item_list/ScrollPositionFetcher.java
package com.gurunars.item_list;

import java.util.List;

class ScrollPositionFetcher {

    int getScrollPosition(int position, Scroller scroller, List items) {
        int upperVisibilityThreshold = scroller.findFirstVisibleItemPosition() + 3;
        int lowerVisibilityThreshold = scroller.findLastVisibleItemPosition() - 3;

        if (upperVisibilityThreshold < position && position < lowerVisibilityThreshold) {
            return -1;  // No change
        }

        position += position <= upperVisibilityThreshold ? -3 : +3;
        return Math.min(Math.max(position, 0), items.size()-1);
    }

}

File: item-list/src/main/java/com/gurunars/item_list/ScrollPositionFetcher.kt
package com.gurunars.item_list

internal class ScrollPositionFetcher {

    fun getScrollPosition(position: Int, scroller: Scroller, items: List<*>): Int {
        var newPosition = position
        val upperVisibilityThreshold = scroller.findFirstVisibleItemPosition() + 3
        val lowerVisibilityThreshold = scroller.findLastVisibleItemPosition() - 3

        if (upperVisibilityThreshold < newPosition && newPosition < lowerVisibilityThreshold) {
            return -1  // No change
        }

        newPosition += if (newPosition <= upperVisibilityThreshold) -3 else +3
        return Math.min(Math.max(newPosition, 0), items.size - 1)
    }

}
